persistence/pre_persistence/account.go

```diff
@@ -0,0 +1,357 @@
+		return err
+	}
+	err = codec.Unmarshal(val, &a)
+	if err != nil {
+		return err
+	}
+	s, err := StringToBigInt(a.Amount)
+	if err != nil {
+		return err
+	}
+	s2, err := StringToBigInt(amount)
+	if err != nil {
+		return err
+	}
+	if err := op(s, s2); err != nil {
+		return err
+	}
+	a.Amount = BigIntToString(s)
+	bz, err := codec.Marshal(&a)
+	if err != nil {
+		return err
+	}
+	return db.Put(key, bz)
+}
+
+func (m *PrePersistenceContext) AddAccountAmount(address []byte, amount string) error {
+	add := func(s *big.Int, s1 *big.Int) error {
+		s.Add(s, s1)
+		return nil
+	}
+	return m.operationAccountAmount(address, amount, add)
+}
+
+func (m *PrePersistenceContext) SubtractAccountAmount(address []byte, amount string) error {
+	sub := func(s *big.Int, s1 *big.Int) error {
+		s.Sub(s, s1)
+		if s.Sign() == -1 {
+			return types.ErrInsufficientAmountError()
+		}
+		return nil
+	}
+	return m.operationAccountAmount(address, amount, sub)
+}
+
+func (m *PrePersistenceContext) GetAccountAmount(address []byte) (string, error) {
+	codec := GetCodec()
+	account := Account{}
+	db := m.Store()
+	key := append(AccountPrefixKey, address...)
+	val, err := db.Get(key)
+	if err != nil {
+		return EmptyString, err
+	}
+	err = codec.Unmarshal(val, &account)
+	if err != nil {
+		return EmptyString, err
+	}
+	return account.Amount, nil
+}
+
+func (m *PrePersistenceContext) SetAccount(address []byte, amount string) error {
+	codec := GetCodec()
+	account := Account{
+		Address: address,
+		Amount:  amount,
+	}
+	db := m.Store()
+	key := append(AccountPrefixKey, address...)
+	bz, err := codec.Marshal(&account)
+	if err != nil {
+		return err
+	}
+	return db.Put(key, bz)
+}
+
+func (m *PrePersistenceContext) GetAllAccounts(height int64) (accs []*Account, err error) {
+	codec := GetCodec()
+	accs = make([]*Account, 0)
+	var it iterator.Iterator
+	if height == m.Height {
+		db := m.Store()
+		it = db.NewIterator(&util.Range{
+			Start: AccountPrefixKey,
+			Limit: PrefixEndBytes(AccountPrefixKey),
+		})
+	} else {
+		key := HeightKey(height, AccountPrefixKey)
+		it = m.Parent.GetCommitDB().NewIterator(&util.Range{
+			Start: key,
+			Limit: PrefixEndBytes(key),
+		})
+	}
+	defer it.Release()
+	for valid := it.First(); valid; valid = it.Next() {
+		bz := it.Value()
+		if bytes.Contains(bz, DeletedPrefixKey) {
+			continue
+		}
+		acc := Account{}
+		if err := codec.Unmarshal(bz, &acc); err != nil {
+			return nil, err
+		}
+		accs = append(accs, &acc)
+	}
+	return
+}
+
+func (x *Account) ValidateBasic() types.Error {
+	if x == nil {
+		return types.ErrEmptyAccount()
+	}
+	if x.Address == nil {
+		return types.ErrEmptyAddress()
+	}
+	if len(x.Address) != crypto.AddressLen {
+		return types.ErrInvalidAddressLen(crypto.ErrInvalidAddressLen())
+	}
+	amount := &big.Int{}
+	if _, ok := amount.SetString(x.Amount, 10); !ok {
+		return types.ErrInvalidAmount()
+	}
+	return nil
+}
+
+func (x *Account) SetAddress(address crypto.Address) types.Error {
+	if x == nil {
```

Comment:
How can this ever be nill if we called it on an instance of the struct? Ditto below.